pkg/plugin/stream_reader.go

```diff
@@ -0,0 +1,55 @@
+package plugin
+
+import (
+	"bytes"
+	"io"
+)
+
+// ReceiveFunc is a function that either returns a slice
+// of bytes OR an error. Returning an io.EOF means there
+// is no more data to be read; any other error is considered
+// an actual error.
+type ReceiveFunc func() ([]byte, error)
+
+type CloseSendFunc func() error
+
+type StreamReadCloser struct {
+	buf     *bytes.Buffer
+	receive ReceiveFunc
+	close   CloseSendFunc
+}
+
+func (s *StreamReadCloser) Read(p []byte) (n int, err error) {
+	// if buf exists and holds more than we're trying to read,
+	// read from the buffer
+	if s.buf != nil && s.buf.Len() >= len(p) {
+		return s.buf.Read(p)
+	}
+
+	// if buf is nil, create it
+	if s.buf == nil {
+		s.buf = new(bytes.Buffer)
+	}
+
+	// buf exists but doesn't hold enough data to fill p, so
+	// receive again. If we get an EOF, return what's in the
+	// buffer; else, write the new data to the buffer and
+	// try another read.
+	data, err := s.receive()
+	if err == io.EOF {
+		return s.buf.Read(p)
+	}
+	if err != nil {
+		return 0, err
+	}
+
+	if _, err := s.buf.Write(data); err != nil {
```

Comment:
I wonder if we should copy what we can to `p` (keeping track of `n`), put the leftovers in `s.buf`, and return `n, nil`?